packages/coreutils/src/token.ts

```diff
@@ -34,6 +34,7 @@ export class Token<T> {
    */
   readonly name: string;
 
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
   // @ts-ignore
```

Comment:
This does now look safe to remove as long as we put a linter ignore tag ahead of the class declaration when it complains about the `@typeparam` value `T` never being accessed.

I think we must have had the compiler option `noUnusedParameters` set to `true` when this was originally compiled. With that flag set, this won't compile without the structural property.

I am happy to remove it. Or we could decide to enable `noUnusedParameters`. What do you think?

Reply:
Is it possible to drop the attribute and add an excepted error comment about the unused parameter.
I would prefer keeping `noUnusedParameter` option as it is a very good rule to enforce.

> I'm afak so I cannot try it myself right now.

Reply:
If we tell both the TypeScript compiler and ESLint to ignore the class signature, we can do what you suggest.
```diff
diff --git a/packages/coreutils/src/token.ts b/packages/coreutils/src/token.ts
index 223a676d..3f369494 100644
--- a/packages/coreutils/src/token.ts
+++ b/packages/coreutils/src/token.ts
@@ -15,6 +15,9 @@
  * A token captures the compile-time type of an interface or class in
  * an object which can be used at runtime in a type-safe fashion.
  */
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
 export class Token<T> {
   /**
    * Construct a new token.
@@ -23,7 +26,6 @@ export class Token<T> {
    */
   constructor(name: string) {
     this.name = name;
-    this._tokenStructuralPropertyT = null!;
   }

   /**
@@ -33,8 +35,4 @@ export class Token<T> {
    * This can be useful for debugging and logging.
    */
   readonly name: string;
-
-  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-  // @ts-ignore
-  private _tokenStructuralPropertyT: T;
 }
diff --git a/tsconfigbase.json b/tsconfigbase.json
index 1d3a2bf6..2cc42caf 100644
--- a/tsconfigbase.json
+++ b/tsconfigbase.json
@@ -14,6 +14,7 @@
     "noEmitOnError": true,
     "noImplicitAny": true,
     "noUnusedLocals": true,
+    "noUnusedParameters": true,
     "preserveWatchOutput": true,
     "resolveJsonModule": true,
     "sourceMap": true,
```

Reply:
When I check the rest of the code base, there are hundreds of places where we have unused arguments. Many of them exist for good reasons (a super-class that does not use those arguments but creates the correct signatures). We can update the code to make sure to either set those arguments to begin with an underscore or put the relevant ignore tag on them.

But it does seem rather invasive. I actually went through and changed everything but I suggest we leave things as they are instead.

Reply:
Okay, this is what the diff would look like if we decided to ship with `noUnusedParameters`:

```diff
diff --git a/packages/algorithm/src/sort.ts b/packages/algorithm/src/sort.ts
index a1efdbc4..a6304ba1 100644
--- a/packages/algorithm/src/sort.ts
+++ b/packages/algorithm/src/sort.ts
@@ -45,7 +45,7 @@ export function topologicSort<T>(edges: IterableOrArrayLike<[T, T]>): T[] {
   each(edges, addEdge);
 
   // Visit each node in the graph.
-  graph.forEach((v, k) => {
+  graph.forEach((_, k) => {
     visit(k);
   });
 
diff --git a/packages/application/src/index.ts b/packages/application/src/index.ts
index f9fd94e0..5e3b98a0 100644
--- a/packages/application/src/index.ts
+++ b/packages/application/src/index.ts
@@ -514,7 +514,7 @@ export class Application<T extends Widget> {
    *
    * A subclass may reimplement this method as needed.
    */
-  protected evtResize(event: Event): void {
+  protected evtResize(_: Event): void {
     this.shell.update();
   }
 
diff --git a/packages/commands/src/index.ts b/packages/commands/src/index.ts
index 8e0158b4..6aa25b10 100644
--- a/packages/commands/src/index.ts
+++ b/packages/commands/src/index.ts
@@ -362,10 +362,7 @@ export class CommandRegistry {
    * @returns A boolean indicating whether the command is toggleable,
    *   or `false` if the command is not registered.
    */
-  isToggleable(
-    id: string,
-    args: ReadonlyJSONObject = JSONExt.emptyObject
-  ): boolean {
+  isToggleable(id: string, _: ReadonlyJSONObject): boolean {
     let cmd = this._commands[id];
     return cmd ? cmd.isToggleable : false;
   }
diff --git a/packages/commands/tests/src/index.spec.ts b/packages/commands/tests/src/index.spec.ts
index 254b44d9..f4062659 100644
--- a/packages/commands/tests/src/index.spec.ts
+++ b/packages/commands/tests/src/index.spec.ts
@@ -254,7 +254,7 @@ describe('@lumino/commands', () => {
 
     describe('#icon()', () => {
       const iconRenderer = {
-        render: (host: HTMLElement, options?: any) => {
+        render: (host: HTMLElement, _?: unknown) => {
           const renderNode = document.createElement('div');
           renderNode.className = 'lm-render';
           host.appendChild(renderNode);
@@ -477,7 +477,7 @@ describe('@lumino/commands', () => {
       it('should execute a specific command', () => {
         let called = false;
         let cmd = {
-          execute: (args: JSONObject) => {
+          execute: (_: JSONObject) => {
             called = true;
           }
         };
@@ -501,7 +501,7 @@ describe('@lumino/commands', () => {
 
       it('should reject if the command throws an error', done => {
         let cmd = {
-          execute: (args: JSONObject) => {
+          execute: (_: JSONObject) => {
             throw new Error('');
           }
         };
@@ -585,7 +585,7 @@ describe('@lumino/commands', () => {
       it('should emit a key binding changed signal when added and removed', () => {
         let added = false;
         registry.addCommand('test', { execute: () => {} });
-        registry.keyBindingChanged.connect((sender, args) => {
+        registry.keyBindingChanged.connect((_, args) => {
           added = args.type === 'added';
         });
         let binding = registry.addKeyBinding({
diff --git a/packages/coreutils/src/token.ts b/packages/coreutils/src/token.ts
index 223a676d..3f369494 100644
--- a/packages/coreutils/src/token.ts
+++ b/packages/coreutils/src/token.ts
@@ -15,6 +15,9 @@
  * A token captures the compile-time type of an interface or class in
  * an object which can be used at runtime in a type-safe fashion.
  */
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
 export class Token<T> {
   /**
    * Construct a new token.
@@ -23,7 +26,6 @@ export class Token<T> {
    */
   constructor(name: string) {
     this.name = name;
-    this._tokenStructuralPropertyT = null!;
   }
 
   /**
@@ -33,8 +35,4 @@ export class Token<T> {
    * This can be useful for debugging and logging.
    */
   readonly name: string;
-
-  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-  // @ts-ignore
-  private _tokenStructuralPropertyT: T;
 }
diff --git a/packages/datagrid/src/basickeyhandler.ts b/packages/datagrid/src/basickeyhandler.ts
index 64d7a519..36c0f9ab 100644
--- a/packages/datagrid/src/basickeyhandler.ts
+++ b/packages/datagrid/src/basickeyhandler.ts
@@ -738,7 +738,7 @@ export class BasicKeyHandler implements DataGrid.IKeyHandler {
    *
    * @param event - The keyboard event of interest.
    */
-  protected onEscape(grid: DataGrid, event: KeyboardEvent): void {
+  protected onEscape(grid: DataGrid, _: KeyboardEvent): void {
     if (grid.selectionModel) {
       grid.selectionModel.clear();
     }
@@ -751,7 +751,7 @@ export class BasicKeyHandler implements DataGrid.IKeyHandler {
    *
    * @param event - The keyboard event of interest.
    */
-  protected onDelete(grid: DataGrid, event: KeyboardEvent): void {
+  protected onDelete(grid: DataGrid, _: KeyboardEvent): void {
     if (grid.editable && !grid.selectionModel!.isEmpty) {
       const dataModel = grid.dataModel as MutableDataModel;
       // Fetch the max row and column.
diff --git a/packages/datagrid/src/basicmousehandler.ts b/packages/datagrid/src/basicmousehandler.ts
index 61fdde9d..75969885 100644
--- a/packages/datagrid/src/basicmousehandler.ts
+++ b/packages/datagrid/src/basicmousehandler.ts
@@ -119,7 +119,7 @@ export class BasicMouseHandler implements DataGrid.IMouseHandler {
    *
    * @param event - The mouse hover event of interest.
    */
-  onMouseLeave(grid: DataGrid, event: MouseEvent): void {
+  onMouseLeave(grid: DataGrid, _: MouseEvent): void {
     // TODO support user-defined hover popups.
 
     // Clear the viewport cursor.
@@ -590,7 +590,7 @@ export class BasicMouseHandler implements DataGrid.IMouseHandler {
    *
    * @param event - The mouse up event of interest.
    */
-  onMouseUp(grid: DataGrid, event: MouseEvent): void {
+  onMouseUp(_g: DataGrid, _e: MouseEvent): void {
     this.release();
   }
 
@@ -642,7 +642,7 @@ export class BasicMouseHandler implements DataGrid.IMouseHandler {
    *
    * @param event - The context menu event of interest.
    */
-  onContextMenu(grid: DataGrid, event: MouseEvent): void {
+  onContextMenu(_g: DataGrid, _e: MouseEvent): void {
     // TODO support user-defined context menus
   }
 
diff --git a/packages/datagrid/src/celleditor.ts b/packages/datagrid/src/celleditor.ts
index 50ebe519..375fcae5 100644
--- a/packages/datagrid/src/celleditor.ts
+++ b/packages/datagrid/src/celleditor.ts
@@ -148,8 +148,8 @@ export class PassInputValidator implements ICellInputValidator {
    * @returns An object with validation result.
    */
   validate(
-    cell: CellEditor.CellConfig,
-    value: unknown
+    _c: CellEditor.CellConfig,
+    _v: unknown
   ): ICellInputValidatorResponse {
     return { valid: true };
   }
@@ -169,7 +169,7 @@ export class TextInputValidator implements ICellInputValidator {
    * @returns An object with validation result.
    */
   validate(
-    cell: CellEditor.CellConfig,
+    _: CellEditor.CellConfig,
     value: string
   ): ICellInputValidatorResponse {
     if (value === null) {
@@ -241,7 +241,7 @@ export class IntegerInputValidator implements ICellInputValidator {
    * @returns An object with validation result.
    */
   validate(
-    cell: CellEditor.CellConfig,
+    _: CellEditor.CellConfig,
     value: number
   ): ICellInputValidatorResponse {
     if (value === null) {
@@ -300,7 +300,7 @@ export class NumberInputValidator implements ICellInputValidator {
    * @returns An object with validation result.
    */
   validate(
-    cell: CellEditor.CellConfig,
+    _: CellEditor.CellConfig,
     value: number
   ): ICellInputValidatorResponse {
     if (value === null) {
@@ -721,12 +721,12 @@ export abstract class CellEditor implements ICellEditor, IDisposable {
     this.viewportOccluder.appendChild(this.editorContainer);
 
     // update mouse event pass-through state based on input validity
-    this.editorContainer.addEventListener('mouseleave', (event: MouseEvent) => {
+    this.editorContainer.addEventListener('mouseleave', (_: MouseEvent) => {
       this.viewportOccluder.style.pointerEvents = this._validInput
         ? 'none'
         : 'auto';
     });
-    this.editorContainer.addEventListener('mouseenter', (event: MouseEvent) => {
+    this.editorContainer.addEventListener('mouseenter', (_: MouseEvent) => {
       this.viewportOccluder.style.pointerEvents = 'none';
     });
   }
@@ -829,7 +829,7 @@ export abstract class InputCellEditor extends CellEditor {
   /**
    * Start editing the cell.
    */
-  protected startEditing() {
+  protected startEditing(): void {
     this.createWidget();
 
     const cell = this.cell;
@@ -842,15 +842,15 @@ export abstract class InputCellEditor extends CellEditor {
     this.bindEvents();
   }
 
-  protected deserialize(value: any): any {
+  protected deserialize(value: unknown): any {
     if (value === null || value === undefined) {
       return '';
     }
 
-    return value.toString();
+    return (value as any).toString();
   }
 
-  protected createWidget() {
+  protected createWidget(): void {
     const input = document.createElement('input');
     input.classList.add('lm-DataGrid-cellEditorWidget');
     input.classList.add('lm-DataGrid-cellEditorInput');
@@ -860,7 +860,7 @@ export abstract class InputCellEditor extends CellEditor {
     this.input = input;
   }
 
-  protected bindEvents() {
+  protected bindEvents(): void {
     this.input.addEventListener('keydown', this);
     this.input.addEventListener('blur', this);
     this.input.addEventListener('input', this);
@@ -902,7 +902,7 @@ export abstract class InputCellEditor extends CellEditor {
     }
   }
 
-  private _onInput(event: Event) {
+  private _onInput(_: Event) {
     this.inputChanged.emit(void 0);
   }
 
@@ -931,7 +931,7 @@ export class NumberCellEditor extends InputCellEditor {
   /**
    * Start editing the cell.
    */
-  protected startEditing() {
+  protected startEditing():void {
     super.startEditing();
 
     this.input.step = 'any';
@@ -982,7 +982,7 @@ export class IntegerCellEditor extends InputCellEditor {
   /**
    * Start editing the cell.
    */
-  protected startEditing() {
+  protected startEditing():void {
     super.startEditing();
 
     this.input.step = '1';
@@ -1049,7 +1049,7 @@ export class DateCellEditor extends CellEditor {
   /**
    * Dispose of the resources held by cell editor.
    */
-  dispose() {
+  dispose(): void {
     if (this.isDisposed) {
       return;
     }
@@ -1062,7 +1062,7 @@ export class DateCellEditor extends CellEditor {
   /**
    * Start editing the cell.
    */
-  protected startEditing() {
+  protected startEditing(): void {
     this._createWidget();
 
     const cell = this.cell;
@@ -1171,7 +1171,7 @@ export class BooleanCellEditor extends CellEditor {
   /**
    * Dispose of the resources held by cell editor.
    */
-  dispose() {
+  dispose(): void {
     if (this.isDisposed) {
       return;
     }
@@ -1184,7 +1184,7 @@ export class BooleanCellEditor extends CellEditor {
   /**
    * Start editing the cell.
    */
-  protected startEditing() {
+  protected startEditing(): void {
     this._createWidget();
 
     const cell = this.cell;
@@ -1277,7 +1277,7 @@ export class OptionCellEditor extends CellEditor {
   /**
    * Dispose of the resources held by cell editor.
    */
-  dispose() {
+  dispose(): void {
     if (this.isDisposed) {
       return;
     }
@@ -1292,7 +1292,7 @@ export class OptionCellEditor extends CellEditor {
   /**
    * Start editing the cell.
    */
-  protected startEditing() {
+  protected startEditing(): void {
     const cell = this.cell;
     const cellInfo = this.getCellInfo(cell);
     const metadata = cell.grid.dataModel!.metadata(
@@ -1463,7 +1463,7 @@ export class DynamicOptionCellEditor extends CellEditor {
   /**
    * Dispose of the resources held by cell editor.
    */
-  dispose() {
+  dispose(): void {
     if (this.isDisposed) {
       return;
     }
@@ -1476,7 +1476,7 @@ export class DynamicOptionCellEditor extends CellEditor {
   /**
    * Start editing the cell.
    */
-  protected startEditing() {
+  protected startEditing(): void {
     this._createWidget();
 
     const cell = this.cell;
diff --git a/packages/datagrid/src/datagrid.ts b/packages/datagrid/src/datagrid.ts
index 7dda9b21..2283c354 100644
--- a/packages/datagrid/src/datagrid.ts
+++ b/packages/datagrid/src/datagrid.ts
@@ -2018,14 +2018,14 @@ export class DataGrid extends Widget {
   /**
    * A message handler invoked on an `'activate-request'` message.
    */
-  protected onActivateRequest(msg: Message): void {
+  protected onActivateRequest(_: Message): void {
     this.viewport.node.focus({ preventScroll: true });
   }
 
   /**
    * A message handler invoked on a `'before-attach'` message.
    */
-  protected onBeforeAttach(msg: Message): void {
+  protected onBeforeAttach(_: Message): void {
     window.addEventListener('resize', this);
     this.node.addEventListener('wheel', this);
     this._viewport.node.addEventListener('keydown', this);
@@ -2041,7 +2041,7 @@ export class DataGrid extends Widget {
   /**
    * A message handler invoked on an `'after-detach'` message.
    */
-  protected onAfterDetach(msg: Message): void {
+  protected onAfterDetach(_: Message): void {
     window.removeEventListener('resize', this);
     this.node.removeEventListener('wheel', this);
     this._viewport.node.removeEventListener('keydown', this);
@@ -2056,7 +2056,7 @@ export class DataGrid extends Widget {
   /**
    * A message handler invoked on a `'before-show'` message.
    */
-  protected onBeforeShow(msg: Message): void {
+  protected onBeforeShow(_: Message): void {
     this.repaintContent();
     this.repaintOverlay();
   }
@@ -2064,7 +2064,7 @@ export class DataGrid extends Widget {
   /**
    * A message handler invoked on a `'resize'` message.
    */
-  protected onResize(msg: Widget.ResizeMessage): void {
+  protected onResize(_: Widget.ResizeMessage): void {
     if (this._editorController) {
       this._editorController.cancel();
     }
@@ -2378,7 +2378,7 @@ export class DataGrid extends Widget {
   /**
    * A message hook invoked on a viewport `'scroll-request'` message.
    */
-  private _onViewportScrollRequest(msg: Message): void {
+  private _onViewportScrollRequest(_: Message): void {
     this._scrollTo(this._hScrollBar.value, this._vScrollBar.value);
   }
 
@@ -2495,7 +2495,7 @@ export class DataGrid extends Widget {
   /**
    * A message hook invoked on a viewport `'overlay-paint-request'` message.
    */
-  private _onViewportOverlayPaintRequest(msg: Message): void {
+  private _onViewportOverlayPaintRequest(_: Message): void {
     // Bail early if the viewport is not visible.
     if (!this._viewport.isVisible) {
       return;
@@ -2537,7 +2537,7 @@ export class DataGrid extends Widget {
   /**
    * Handle the `thumbMoved` signal from a scroll bar.
    */
-  private _onThumbMoved(sender: ScrollBar): void {
+  private _onThumbMoved(_: ScrollBar): void {
     MessageLoop.postMessage(this._viewport, Private.ScrollRequest);
   }
 
@@ -2573,7 +2573,7 @@ export class DataGrid extends Widget {
    * A signal handler for the data model `changed` signal.
    */
   private _onDataModelChanged(
-    sender: DataModel,
+    _: DataModel,
     args: DataModel.ChangedArgs
   ): void {
     switch (args.type) {
@@ -2609,7 +2609,7 @@ export class DataGrid extends Widget {
   /**
    * A signal handler for the selection model `changed` signal.
    */
-  private _onSelectionsChanged(sender: SelectionModel): void {
+  private _onSelectionsChanged(_: SelectionModel): void {
     this.repaintOverlay();
   }
 
@@ -2892,7 +2892,7 @@ export class DataGrid extends Widget {
   /**
    * Handle a full data model reset.
    */
-  private _onModelReset(args: DataModel.ModelResetArgs): void {
+  private _onModelReset(_: DataModel.ModelResetArgs): void {
     // Look up the various current section counts.
     let nr = this._rowSections.count;
     let nc = this._columnSections.count;
diff --git a/packages/datagrid/src/datamodel.ts b/packages/datagrid/src/datamodel.ts
index 5cf81796..3907030e 100644
--- a/packages/datagrid/src/datamodel.ts
+++ b/packages/datagrid/src/datamodel.ts
@@ -72,7 +72,7 @@ export abstract class DataModel {
    * cell region.
    * @param region the target cell region.
    */
-  groupCount(region: DataModel.CellRegion): number {
+  groupCount(_: DataModel.CellRegion): number {
     return 0;
   }
 
@@ -95,9 +95,9 @@ export abstract class DataModel {
    * The default implementation returns `{}`.
    */
   metadata(
-    region: DataModel.CellRegion,
-    row: number,
-    column: number
+    _: DataModel.CellRegion,
+    _r: number,
+    _c: number
   ): DataModel.Metadata {
     return DataModel.emptyMetadata;
   }
@@ -108,7 +108,7 @@ export abstract class DataModel {
    * @param groupIndex the group index of the cell group.
    * @returns a cell group.
    */
-  group(region: DataModel.CellRegion, groupIndex: number): CellGroup | null {
+  group(_r: DataModel.CellRegion, _g: number): CellGroup | null {
     return null;
   }
 
@@ -150,7 +150,7 @@ export abstract class MutableDataModel extends DataModel {
     region: DataModel.CellRegion,
     row: number,
     column: number,
-    value: any
+    value: unknown
   ): boolean;
 }
 
diff --git a/packages/datagrid/src/jsonmodel.ts b/packages/datagrid/src/jsonmodel.ts
index 81e05b09..fb321c96 100644
--- a/packages/datagrid/src/jsonmodel.ts
+++ b/packages/datagrid/src/jsonmodel.ts
@@ -121,7 +121,7 @@ export class JSONModel extends DataModel {
    */
   metadata(
     region: DataModel.CellRegion,
-    row: number,
+    _: number,
     column: number
   ): DataModel.Metadata {
     if (region === 'body' || region === 'column-header') {
diff --git a/packages/datagrid/src/notification.ts b/packages/datagrid/src/notification.ts
index 330ab48a..7db26c42 100644
--- a/packages/datagrid/src/notification.ts
+++ b/packages/datagrid/src/notification.ts
@@ -121,7 +121,7 @@ export class Notification extends Widget {
   /**
    * A method invoked on a 'before-attach' message.
    */
-  protected onBeforeAttach(msg: Message): void {
+  protected onBeforeAttach(_: Message): void {
     this.node.addEventListener('mousedown', this);
     this.update();
   }
@@ -129,14 +129,14 @@ export class Notification extends Widget {
   /**
    * A method invoked on an 'after-detach' message.
    */
-  protected onAfterDetach(msg: Message): void {
+  protected onAfterDetach(_: Message): void {
     this.node.removeEventListener('mousedown', this);
   }
 
   /**
    * A method invoked on an 'update-request' message.
    */
-  protected onUpdateRequest(msg: Message): void {
+  protected onUpdateRequest(_: Message): void {
     const targetRect = this._target.getBoundingClientRect();
     const style = this.node.style;
 
diff --git a/packages/datagrid/src/selectionmodel.ts b/packages/datagrid/src/selectionmodel.ts
index 912275b8..7ea37ce2 100644
--- a/packages/datagrid/src/selectionmodel.ts
+++ b/packages/datagrid/src/selectionmodel.ts
@@ -189,10 +189,7 @@ export abstract class SelectionModel {
    *
    * The default implementation of this method is a no-op.
    */
-  protected onDataModelChanged(
-    sender: DataModel,
-    args: DataModel.ChangedArgs
-  ): void {
+  protected onDataModelChanged(_s: DataModel, _a: DataModel.ChangedArgs): void {
     // pass
   }
 
diff --git a/packages/dragdrop/src/index.ts b/packages/dragdrop/src/index.ts
index e093f0fb..c07e2770 100644
--- a/packages/dragdrop/src/index.ts
+++ b/packages/dragdrop/src/index.ts
@@ -239,7 +239,7 @@ export class Drag implements IDisposable {
     this._attachDragImage(clientX, clientY);
 
     // Create the promise which will be resolved on completion.
-    this._promise = new Promise<DropAction>((resolve, reject) => {
+    this._promise = new Promise<DropAction>(resolve => {
       this._resolve = resolve;
     });
 
diff --git a/packages/messaging/src/index.ts b/packages/messaging/src/index.ts
index d72a6812..f7035a36 100644
--- a/packages/messaging/src/index.ts
+++ b/packages/messaging/src/index.ts
@@ -91,7 +91,7 @@ export class Message {
    *
    * The default implementation always returns `false`.
    */
-  conflate(other: Message): boolean {
+  conflate(_: Message): boolean {
     return false;
   }
 }
@@ -126,7 +126,7 @@ export class ConflatableMessage extends Message {
    * #### Notes
    * This method always returns `true`.
    */
-  conflate(other: ConflatableMessage): boolean {
+  conflate(_: ConflatableMessage): boolean {
     return true;
   }
 }
diff --git a/packages/messaging/tests/src/index.spec.ts b/packages/messaging/tests/src/index.spec.ts
index dc091a2c..d60b9917 100644
--- a/packages/messaging/tests/src/index.spec.ts
+++ b/packages/messaging/tests/src/index.spec.ts
@@ -27,7 +27,7 @@ class Handler implements IMessageHandler {
 }
 
 class BadHandler implements IMessageHandler {
-  processMessage(msg: Message): void {
+  processMessage(_: Message): void {
     throw new Error('process error');
   }
 }
diff --git a/packages/polling/src/index.ts b/packages/polling/src/index.ts
index af44b6c8..04398462 100644
--- a/packages/polling/src/index.ts
+++ b/packages/polling/src/index.ts
@@ -162,6 +162,8 @@ export namespace IPoll {
  *
  * @typeparam V - Arguments for the underlying function. Defaults to any[].
  */
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
 export interface IRateLimiter<T = any, U = any, V extends any[] = any[]>
   extends IDisposable {
diff --git a/packages/polling/tests/src/poll.spec.ts b/packages/polling/tests/src/poll.spec.ts
index 04abac5c..84ff9ab8 100644
--- a/packages/polling/tests/src/poll.spec.ts
+++ b/packages/polling/tests/src/poll.spec.ts
@@ -9,7 +9,7 @@ import { IPoll, Poll } from '@lumino/polling';
  * Return a promise that resolves in the given milliseconds with the given value.
  */
 function sleep<T>(milliseconds: number = 0, value?: T): Promise<T | undefined> {
-  return new Promise((resolve, reject) => {
+  return new Promise((resolve) => {
     setTimeout(() => {
       resolve(value);
     }, milliseconds);
diff --git a/packages/properties/src/index.ts b/packages/properties/src/index.ts
index a40329a0..6095782a 100644
--- a/packages/properties/src/index.ts
+++ b/packages/properties/src/index.ts
@@ -222,7 +222,7 @@ export namespace AttachedProperty {
    * This will clear all property values for the owner, but it will
    * **not** run the change notification for any of the properties.
    */
-  export function clearData(owner: any): void {
+  export function clearData(owner: unknown): void {
     Private.ownerData.delete(owner);
   }
 }
@@ -258,7 +258,7 @@ namespace Private {
    *
    * This will create the map if one does not already exist.
    */
-  export function ensureMap(owner: any): PropertyMap {
+  export function ensureMap(owner: unknown): PropertyMap {
     let map = ownerData.get(owner);
     if (map) {
       return map;
diff --git a/packages/properties/tests/src/index.spec.ts b/packages/properties/tests/src/index.spec.ts
index 85ca90ec..f0efbb3c 100644
--- a/packages/properties/tests/src/index.spec.ts
+++ b/packages/properties/tests/src/index.spec.ts
@@ -1,4 +1,3 @@
-/* eslint-disable @typescript-eslint/no-empty-function */
 // Copyright (c) Jupyter Development Team.
 // Distributed under the terms of the Modified BSD License.
 /*-----------------------------------------------------------------------------
@@ -22,10 +21,10 @@ describe('@lumino/properties', () => {
       it('should accept a single options argument', () => {
         let p = new AttachedProperty<Model, number>({
           name: 'p',
-          create: owner => 42,
-          coerce: (owner, value) => Math.max(0, value),
+          create: _ => 42,
+          coerce: (_, value) => Math.max(0, value),
           compare: (oldValue, newValue) => oldValue === newValue,
-          changed: (owner, oldValue, newValue) => {}
+          changed: (_, _o, _n) => { return; }
         });
         expect(p).to.be.an.instanceof(AttachedProperty);
       });
@@ -63,7 +62,7 @@ describe('@lumino/properties', () => {
       it('should not invoke the coerce function', () => {
         let called = false;
         let create = () => 0;
-        let coerce = (m: Model, v: number) => ((called = true), v);
+        let coerce = (_: Model, v: number) => ((called = true), v);
         let p1 = new AttachedProperty<Model, number>({
           name: 'p1',
           create,
@@ -197,7 +196,7 @@ describe('@lumino/properties', () => {
       it('should invoke the changed function if the value changes', () => {
         let oldvals: number[] = [];
         let newvals: number[] = [];
-        let changed = (m: Model, o: number, n: number) => {
+        let changed = (_: Model, o: number, n: number) => {
           oldvals.push(o);
           newvals.push(n);
         };
@@ -235,7 +234,7 @@ describe('@lumino/properties', () => {
 
       it('should invoke the coerce function on the new value', () => {
         let create = () => 0;
-        let coerce = (o: Model, v: number) => Math.max(0, v);
+        let coerce = (_: Model, v: number) => Math.max(0, v);
         let p = new AttachedProperty<Model, number>({
           name: 'p',
           create,
@@ -271,7 +270,7 @@ describe('@lumino/properties', () => {
       it('should invoke the compare function if there is a changed function', () => {
         let called = false;
         let create = () => 0;
-        let changed = () => {};
+        let changed = () => undefined;
         let compare = (v1: number, v2: number) => ((called = true), v1 === v2);
         let p = new AttachedProperty<Model, number>({
           name: 'p',
@@ -290,7 +289,7 @@ describe('@lumino/properties', () => {
         let changed = () => {
           called = true;
         };
-        let compare = (v1: number, v2: number) => true;
+        let compare = (_1: number, _2: number) => true;
         let p1 = new AttachedProperty<Model, number>({
           name: 'p1',
           create,
@@ -318,7 +317,7 @@ describe('@lumino/properties', () => {
         let min = 20;
         let max = 50;
         let create = () => 0;
-        let coerce = (m: Model, v: number) => Math.max(min, Math.min(v, max));
+        let coerce = (_: Model, v: number) => Math.max(min, Math.min(v, max));
         let p = new AttachedProperty<Model, number>({
           name: 'p',
           create,
@@ -339,7 +338,7 @@ describe('@lumino/properties', () => {
       it('should invoke the changed function if the value changes', () => {
         let called = false;
         let create = () => 0;
-        let coerce = (m: Model, v: number) => Math.max(20, v);
+        let coerce = (_: Model, v: number) => Math.max(20, v);
         let changed = () => {
           called = true;
         };
@@ -358,8 +357,8 @@ describe('@lumino/properties', () => {
         let oldval = -1;
         let newval = -1;
         let create = () => 0;
-        let coerce = (m: Model, v: number) => Math.max(20, v);
-        let changed = (m: Model, o: number, n: number) => {
+        let coerce = (_: Model, v: number) => Math.max(20, v);
+        let changed = (_: Model, o: number, n: number) => {
           oldval = o;
           newval = n;
         };
diff --git a/packages/signaling/tests/src/index.spec.ts b/packages/signaling/tests/src/index.spec.ts
index 9b7f4734..ab16a458 100644
--- a/packages/signaling/tests/src/index.spec.ts
+++ b/packages/signaling/tests/src/index.spec.ts
@@ -45,7 +45,7 @@ class TestHandler {
     this.twoValue = args;
   }
 
-  onThree(sender: TestObject, args: string[]): void {
+  onThree(_: TestObject, args: string[]): void {
     args.push(this.name);
   }
 
diff --git a/packages/virtualdom/tests/src/index.spec.ts b/packages/virtualdom/tests/src/index.spec.ts
index 596a53e1..8ebaa378 100644
--- a/packages/virtualdom/tests/src/index.spec.ts
+++ b/packages/virtualdom/tests/src/index.spec.ts
@@ -77,8 +77,8 @@ describe('@lumino/virtualdom', () => {
 
   describe('VirtualElement with custom .renderer', () => {
     let mockRenderer = {
-      render: (host: HTMLElement) => {},
-      unrender: (host: HTMLElement) => {}
+      render: (_: HTMLElement) => {},
+      unrender: (_: HTMLElement) => {}
     };
 
     describe('#constructor()', () => {
@@ -309,8 +309,8 @@ describe('@lumino/virtualdom', () => {
     let tag = 'div';
     let attrs = { className: 'baz' };
     let mockRenderer = {
-      render: (host: HTMLElement) => {},
-      unrender: (host: HTMLElement) => {}
+      render: (_: HTMLElement) => {},
+      unrender: (_: HTMLElement) => {}
     };
 
     it('should create a new virtual element with custom renderer', () => {
diff --git a/packages/widgets/src/accordionpanel.ts b/packages/widgets/src/accordionpanel.ts
index 20a25e31..831c83e0 100644
--- a/packages/widgets/src/accordionpanel.ts
+++ b/packages/widgets/src/accordionpanel.ts
@@ -344,11 +344,9 @@ export namespace AccordionPanel {
     /**
      * Render the collapse indicator for a section title.
      *
-     * @param data - The data to use for rendering the section title.
-     *
      * @returns A element representing the collapse indicator.
      */
-    createCollapseIcon(data: Title<Widget>): HTMLElement {
+    createCollapseIcon(_: Title<Widget>): HTMLElement {
       return document.createElement('span');
     }
 
diff --git a/packages/widgets/src/boxlayout.ts b/packages/widgets/src/boxlayout.ts
index 558c0580..ff3098d7 100644
--- a/packages/widgets/src/boxlayout.ts
+++ b/packages/widgets/src/boxlayout.ts
@@ -201,7 +201,7 @@ export class BoxLayout extends PanelLayout {
   protected moveWidget(
     fromIndex: number,
     toIndex: number,
-    widget: Widget
+    _: Widget
   ): void {
     // Move the layout item for the widget.
     ArrayExt.move(this._items, fromIndex, toIndex);
@@ -269,14 +269,14 @@ export class BoxLayout extends PanelLayout {
   /**
    * A message handler invoked on a `'child-shown'` message.
    */
-  protected onChildShown(msg: Widget.ChildMessage): void {
+  protected onChildShown(_: Widget.ChildMessage): void {
     this.parent!.fit();
   }
 
   /**
    * A message handler invoked on a `'child-hidden'` message.
    */
-  protected onChildHidden(msg: Widget.ChildMessage): void {
+  protected onChildHidden(_: Widget.ChildMessage): void {
     this.parent!.fit();
   }
 
@@ -292,7 +292,7 @@ export class BoxLayout extends PanelLayout {
   /**
    * A message handler invoked on an `'update-request'` message.
    */
-  protected onUpdateRequest(msg: Message): void {
+  protected onUpdateRequest(_: Message): void {
     if (this.parent!.isVisible) {
       this._update(-1, -1);
     }
@@ -301,7 +301,7 @@ export class BoxLayout extends PanelLayout {
   /**
    * A message handler invoked on a `'fit-request'` message.
    */
-  protected onFitRequest(msg: Message): void {
+  protected onFitRequest(_: Message): void {
     if (this.parent!.isAttached) {
       this._fit();
     }
@@ -617,7 +617,7 @@ namespace Private {
   export const stretchProperty = new AttachedProperty<Widget, number>({
     name: 'stretch',
     create: () => 0,
-    coerce: (owner, value) => Math.max(0, Math.floor(value)),
+    coerce: (_, value) => Math.max(0, Math.floor(value)),
     changed: onChildSizingChanged
   });
 
@@ -627,7 +627,7 @@ namespace Private {
   export const sizeBasisProperty = new AttachedProperty<Widget, number>({
     name: 'sizeBasis',
     create: () => 0,
-    coerce: (owner, value) => Math.max(0, Math.floor(value)),
+    coerce: (_, value) => Math.max(0, Math.floor(value)),
     changed: onChildSizingChanged
   });
 
diff --git a/packages/widgets/src/commandpalette.ts b/packages/widgets/src/commandpalette.ts
index 6e77864a..26106b9a 100644
--- a/packages/widgets/src/commandpalette.ts
+++ b/packages/widgets/src/commandpalette.ts
@@ -253,7 +253,7 @@ export class CommandPalette extends Widget {
   /**
    * A message handler invoked on a `'before-attach'` message.
    */
-  protected onBeforeAttach(msg: Message): void {
+  protected onBeforeAttach(_: Message): void {
     this.node.addEventListener('click', this);
     this.node.addEventListener('keydown', this);
     this.node.addEventListener('input', this);
@@ -264,7 +264,7 @@ export class CommandPalette extends Widget {
   /**
    * A message handler invoked on an `'after-detach'` message.
    */
-  protected onAfterDetach(msg: Message): void {
+  protected onAfterDetach(_: Message): void {
     this.node.removeEventListener('click', this);
     this.node.removeEventListener('keydown', this);
     this.node.removeEventListener('input', this);
@@ -275,7 +275,7 @@ export class CommandPalette extends Widget {
   /**
    * A message handler invoked on an `'activate-request'` message.
    */
-  protected onActivateRequest(msg: Message): void {
+  protected onActivateRequest(_: Message): void {
     if (this.isAttached) {
       let input = this.inputNode;
       input.focus();
@@ -286,7 +286,7 @@ export class CommandPalette extends Widget {
   /**
    * A message handler invoked on an `'update-request'` message.
    */
-  protected onUpdateRequest(msg: Message): void {
+  protected onUpdateRequest(_: Message): void {
     // Fetch the current query text and content node.
     let query = this.inputNode.value;
     let contentNode = this.contentNode;
diff --git a/packages/widgets/src/docklayout.ts b/packages/widgets/src/docklayout.ts
index ec7b2647..f9285252 100644
--- a/packages/widgets/src/docklayout.ts
+++ b/packages/widgets/src/docklayout.ts
@@ -343,9 +343,8 @@ export class DockLayout extends Layout {
       this._root = Private.realizeAreaConfig(
         mainConfig,
         {
-          // Ignoring optional `document` argument as we must reuse `this._document`
-          createTabBar: (document?: Document | ShadowRoot) =>
-            this._createTabBar(),
+          // Ignore `document` argument because we must reuse `this._document`.
+          createTabBar: (_?: Document | ShadowRoot) => this._createTabBar(),
           createHandle: () => this._createHandle()
         },
         this._document
@@ -618,14 +617,14 @@ export class DockLayout extends Layout {
   /**
    * A message handler invoked on a `'child-shown'` message.
    */
-  protected onChildShown(msg: Widget.ChildMessage): void {
+  protected onChildShown(_: Widget.ChildMessage): void {
     this.parent!.fit();
   }
 
   /**
    * A message handler invoked on a `'child-hidden'` message.
    */
-  protected onChildHidden(msg: Widget.ChildMessage): void {
+  protected onChildHidden(_: Widget.ChildMessage): void {
     this.parent!.fit();
   }
 
@@ -641,7 +640,7 @@ export class DockLayout extends Layout {
   /**
    * A message handler invoked on an `'update-request'` message.
    */
-  protected onUpdateRequest(msg: Message): void {
+  protected onUpdateRequest(_: Message): void {
     if (this.parent!.isVisible) {
       this._update(-1, -1);
     }
@@ -650,7 +649,7 @@ export class DockLayout extends Layout {
   /**
    * A message handler invoked on a `'fit-request'` message.
    */
-  protected onFitRequest(msg: Message): void {
+  protected onFitRequest(_: Message): void {
     if (this.parent!.isAttached) {
       this._fit();
     }
@@ -1605,7 +1604,7 @@ namespace Private {
      * Find the split layout node which contains the given handle.
      */
     findSplitNode(
-      handle: HTMLDivElement
+      _: HTMLDivElement
     ): { index: number; node: SplitLayoutNode } | null {
       return null;
     }
@@ -1651,7 +1650,7 @@ namespace Private {
     /**
      * Fit the layout tree.
      */
-    fit(spacing: number, items: ItemMap): ElementExt.ISizeLimits {
+    fit(_: number, items: ItemMap): ElementExt.ISizeLimits {
       // Set up the limit variables.
       let minWidth = 0;
       let minHeight = 0;
@@ -1712,7 +1711,7 @@ namespace Private {
       top: number,
       width: number,
       height: number,
-      spacing: number,
+      _: number,
       items: ItemMap
     ): void {
       // Update the layout box values.
@@ -1983,7 +1982,7 @@ namespace Private {
 
       // Normalize the sizes based on the sum.
       if (sum === 0) {
-        each(sizes, (size, i) => {
+        each(sizes, (_, i) => {
           sizes[i] = 1 / n;
         });
       } else {
@@ -2081,7 +2080,7 @@ namespace Private {
     }
   }
 
-  export function addAria(widget: Widget, tabBar: TabBar<Widget>) {
+  export function addAria(widget: Widget, tabBar: TabBar<Widget>): void {
     widget.node.setAttribute('role', 'tabpanel');
     let renderer = tabBar.renderer;
     if (renderer instanceof TabBar.Renderer) {
@@ -2094,7 +2093,7 @@ namespace Private {
     }
   }
 
-  export function removeAria(widget: Widget) {
+  export function removeAria(widget: Widget): void {
     widget.node.removeAttribute('role');
     widget.node.removeAttribute('aria-labelledby');
   }
diff --git a/packages/widgets/src/dockpanel.ts b/packages/widgets/src/dockpanel.ts
index ed045d96..43e5f245 100644
--- a/packages/widgets/src/dockpanel.ts
+++ b/packages/widgets/src/dockpanel.ts
@@ -467,7 +467,7 @@ export class DockPanel extends Widget {
   /**
    * A message handler invoked on a `'before-attach'` message.
    */
-  protected onBeforeAttach(msg: Message): void {
+  protected onBeforeAttach(_: Message): void {
     this.node.addEventListener('lm-dragenter', this);
     this.node.addEventListener('lm-dragleave', this);
     this.node.addEventListener('lm-dragover', this);
@@ -478,7 +478,7 @@ export class DockPanel extends Widget {
   /**
    * A message handler invoked on an `'after-detach'` message.
    */
-  protected onAfterDetach(msg: Message): void {
+  protected onAfterDetach(_: Message): void {
     this.node.removeEventListener('lm-dragenter', this);
     this.node.removeEventListener('lm-dragleave', this);
     this.node.removeEventListener('lm-dragover', this);
@@ -876,13 +876,14 @@ export class DockPanel extends Widget {
         right = target!.right;
         bottom = target!.bottom;
         break;
-      case 'widget-tab':
+      case 'widget-tab': {
         const tabHeight = target!.tabBar.node.getBoundingClientRect().height;
         top = target!.top;
         left = target!.left;
         right = target!.right;
         bottom = target!.bottom + target!.height - tabHeight;
         break;
+      }
       default:
         throw 'unreachable';
     }
@@ -947,7 +948,7 @@ export class DockPanel extends Widget {
    * Handle the `currentChanged` signal from a tab bar.
    */
   private _onCurrentChanged(
-    sender: TabBar<Widget>,
+    _: TabBar<Widget>,
     args: TabBar.ICurrentChangedArgs<Widget>
   ): void {
     // Extract the previous and current title from the args.
@@ -983,7 +984,7 @@ export class DockPanel extends Widget {
    * Handle the `tabActivateRequested` signal from a tab bar.
    */
   private _onTabActivateRequested(
-    sender: TabBar<Widget>,
+    _: TabBar<Widget>,
     args: TabBar.ITabActivateRequestedArgs<Widget>
   ): void {
     args.title.owner.activate();
@@ -993,7 +994,7 @@ export class DockPanel extends Widget {
    * Handle the `tabCloseRequested` signal from a tab bar.
    */
   private _onTabCloseRequested(
-    sender: TabBar<Widget>,
+    _: TabBar<Widget>,
     args: TabBar.ITabCloseRequestedArgs<Widget>
   ): void {
     args.title.owner.close();
diff --git a/packages/widgets/src/gridlayout.ts b/packages/widgets/src/gridlayout.ts
index a77f0512..b32826d8 100644
--- a/packages/widgets/src/gridlayout.ts
+++ b/packages/widgets/src/gridlayout.ts
@@ -427,14 +427,14 @@ export class GridLayout extends Layout {
   /**
    * A message handler invoked on a `'child-shown'` message.
    */
-  protected onChildShown(msg: Widget.ChildMessage): void {
+  protected onChildShown(_: Widget.ChildMessage): void {
     this.parent!.fit();
   }
 
   /**
    * A message handler invoked on a `'child-hidden'` message.
    */
-  protected onChildHidden(msg: Widget.ChildMessage): void {
+  protected onChildHidden(_: Widget.ChildMessage): void {
     this.parent!.fit();
   }
 
@@ -450,7 +450,7 @@ export class GridLayout extends Layout {
   /**
    * A message handler invoked on an `'update-request'` message.
    */
-  protected onUpdateRequest(msg: Message): void {
+  protected onUpdateRequest(_: Message): void {
     if (this.parent!.isVisible) {
       this._update(-1, -1);
     }
@@ -459,7 +459,7 @@ export class GridLayout extends Layout {
   /**
    * A message handler invoked on a `'fit-request'` message.
    */
-  protected onFitRequest(msg: Message): void {
+  protected onFitRequest(_: Message): void {
     if (this.parent!.isAttached) {
       this._fit();
     }
diff --git a/packages/widgets/src/layout.ts b/packages/widgets/src/layout.ts
index d4337f9c..cb384f4e 100644
--- a/packages/widgets/src/layout.ts
+++ b/packages/widgets/src/layout.ts
@@ -257,7 +257,7 @@ export abstract class Layout implements IIterable<Widget>, IDisposable {
    *
    * This may be reimplemented by subclasses as needed.
    */
-  protected onResize(msg: Widget.ResizeMessage): void {
+  protected onResize(_: Widget.ResizeMessage): void {
     each(this, widget => {
       MessageLoop.sendMessage(widget, Widget.ResizeMessage.UnknownSize);
     });
@@ -276,7 +276,7 @@ export abstract class Layout implements IIterable<Widget>, IDisposable {
    *
    * This may be reimplemented by subclasses as needed.
    */
-  protected onUpdateRequest(msg: Message): void {
+  protected onUpdateRequest(_: Message): void {
     each(this, widget => {
       MessageLoop.sendMessage(widget, Widget.ResizeMessage.UnknownSize);
     });
@@ -436,7 +436,7 @@ export abstract class Layout implements IIterable<Widget>, IDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onFitRequest(msg: Message): void {}
+  protected onFitRequest(_: Message): void {}
 
   /**
    * A message handler invoked on a `'child-shown'` message.
@@ -444,7 +444,7 @@ export abstract class Layout implements IIterable<Widget>, IDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onChildShown(msg: Widget.ChildMessage): void {}
+  protected onChildShown(_: Widget.ChildMessage): void {}
 
   /**
    * A message handler invoked on a `'child-hidden'` message.
@@ -452,7 +452,7 @@ export abstract class Layout implements IIterable<Widget>, IDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onChildHidden(msg: Widget.ChildMessage): void {}
+  protected onChildHidden(_: Widget.ChildMessage): void {}
 
   private _disposed = false;
   private _fitPolicy: Layout.FitPolicy;
diff --git a/packages/widgets/src/menu.ts b/packages/widgets/src/menu.ts
index b2b11e34..462e5a93 100644
--- a/packages/widgets/src/menu.ts
+++ b/packages/widgets/src/menu.ts
@@ -498,7 +498,7 @@ export class Menu extends Widget {
   /**
    * A message handler invoked on a `'before-attach'` message.
    */
-  protected onBeforeAttach(msg: Message): void {
+  protected onBeforeAttach(_: Message): void {
     this.node.addEventListener('keydown', this);
     this.node.addEventListener('mouseup', this);
     this.node.addEventListener('mousemove', this);
@@ -511,7 +511,7 @@ export class Menu extends Widget {
   /**
    * A message handler invoked on an `'after-detach'` message.
    */
-  protected onAfterDetach(msg: Message): void {
+  protected onAfterDetach(_: Message): void {
     this.node.removeEventListener('keydown', this);
     this.node.removeEventListener('mouseup', this);
     this.node.removeEventListener('mousemove', this);
@@ -524,7 +524,7 @@ export class Menu extends Widget {
   /**
    * A message handler invoked on an `'activate-request'` message.
    */
-  protected onActivateRequest(msg: Message): void {
+  protected onActivateRequest(_: Message): void {
     if (this.isAttached) {
       this.node.focus();
     }
@@ -533,7 +533,7 @@ export class Menu extends Widget {
   /**
    * A message handler invoked on an `'update-request'` message.
    */
-  protected onUpdateRequest(msg: Message): void {
+  protected onUpdateRequest(_: Message): void {
     let items = this._items;
     let renderer = this.renderer;
     let activeIndex = this._activeIndex;
@@ -745,7 +745,7 @@ export class Menu extends Widget {
    * #### Notes
    * This listener is attached to the menu node.
    */
-  private _evtMouseEnter(event: MouseEvent): void {
+  private _evtMouseEnter(_: MouseEvent): void {
     // Synchronize the active ancestor items.
     for (let menu = this._parentMenu; menu; menu = menu._parentMenu) {
       menu._cancelOpenTimer();
@@ -1204,11 +1204,9 @@ export namespace Menu {
     /**
      * Render the submenu icon element for a menu item.
      *
-     * @param data - The data to use for rendering the submenu icon.
-     *
      * @returns A virtual element representing the submenu icon.
      */
-    renderSubmenu(data: IRenderData): VirtualElement {
+    renderSubmenu(_: IRenderData): VirtualElement {
       return h.div({ className: 'lm-Menu-itemSubmenuIcon' });
     }
 
diff --git a/packages/widgets/src/menubar.ts b/packages/widgets/src/menubar.ts
index 52e7fd1e..54c62ba3 100644
--- a/packages/widgets/src/menubar.ts
+++ b/packages/widgets/src/menubar.ts
@@ -353,7 +353,7 @@ export class MenuBar extends Widget {
   /**
    * A message handler invoked on a `'before-attach'` message.
    */
-  protected onBeforeAttach(msg: Message): void {
+  protected onBeforeAttach(_: Message): void {
     this.node.addEventListener('keydown', this);
     this.node.addEventListener('mousedown', this);
     this.node.addEventListener('mousemove', this);
@@ -364,7 +364,7 @@ export class MenuBar extends Widget {
   /**
    * A message handler invoked on an `'after-detach'` message.
    */
-  protected onAfterDetach(msg: Message): void {
+  protected onAfterDetach(_: Message): void {
     this.node.removeEventListener('keydown', this);
     this.node.removeEventListener('mousedown', this);
     this.node.removeEventListener('mousemove', this);
@@ -376,7 +376,7 @@ export class MenuBar extends Widget {
   /**
    * A message handler invoked on an `'activate-request'` message.
    */
-  protected onActivateRequest(msg: Message): void {
+  protected onActivateRequest(_: Message): void {
     if (this.isAttached) {
       this.node.focus();
     }
@@ -385,7 +385,7 @@ export class MenuBar extends Widget {
   /**
    * A message handler invoked on an `'update-request'` message.
    */
-  protected onUpdateRequest(msg: Message): void {
+  protected onUpdateRequest(_: Message): void {
     let menus = this._menus;
     let renderer = this.renderer;
     let activeIndex = this._activeIndex;
@@ -549,7 +549,7 @@ export class MenuBar extends Widget {
   /**
    * Handle the `'mouseleave'` event for the menu bar.
    */
-  private _evtMouseLeave(event: MouseEvent): void {
+  private _evtMouseLeave(_: MouseEvent): void {
     // Reset the active index if there is no open menu.
     if (!this._childMenu) {
       this.activeIndex = -1;
@@ -834,7 +834,7 @@ export namespace MenuBar {
      *
      * @returns The aria attributes object for the item.
      */
-    createItemARIA(data: IRenderData): ElementARIAAttrs {
+    createItemARIA(_: IRenderData): ElementARIAAttrs {
       return { role: 'menuitem', 'aria-haspopup': 'true' };
     }
 
diff --git a/packages/widgets/src/panellayout.ts b/packages/widgets/src/panellayout.ts
index 98197e5c..d49c63b8 100644
--- a/packages/widgets/src/panellayout.ts
+++ b/packages/widgets/src/panellayout.ts
@@ -240,7 +240,7 @@ export class PanelLayout extends Layout {
    * node is moved in the parent's node.
    */
   protected moveWidget(
-    fromIndex: number,
+    _: number,
     toIndex: number,
     widget: Widget
   ): void {
@@ -292,7 +292,7 @@ export class PanelLayout extends Layout {
    * Subclasses may reimplement this method to control how the widget's
    * node is removed from the parent's node.
    */
-  protected detachWidget(index: number, widget: Widget): void {
+  protected detachWidget(_: number, widget: Widget): void {
     // Send a `'before-detach'` message if the parent is attached.
     if (this.parent!.isAttached) {
       MessageLoop.sendMessage(widget, Widget.Msg.BeforeDetach);
diff --git a/packages/widgets/src/scrollbar.ts b/packages/widgets/src/scrollbar.ts
index ea2f65ec..ba00b4ba 100644
--- a/packages/widgets/src/scrollbar.ts
+++ b/packages/widgets/src/scrollbar.ts
@@ -284,7 +284,7 @@ export class ScrollBar extends Widget {
   /**
    * A method invoked on a 'before-attach' message.
    */
-  protected onBeforeAttach(msg: Message): void {
+  protected onBeforeAttach(_: Message): void {
     this.node.addEventListener('mousedown', this);
     this.update();
   }
@@ -292,7 +292,7 @@ export class ScrollBar extends Widget {
   /**
    * A method invoked on an 'after-detach' message.
    */
-  protected onAfterDetach(msg: Message): void {
+  protected onAfterDetach(_: Message): void {
     this.node.removeEventListener('mousedown', this);
     this._releaseMouse();
   }
@@ -300,7 +300,7 @@ export class ScrollBar extends Widget {
   /**
    * A method invoked on an 'update-request' message.
    */
-  protected onUpdateRequest(msg: Message): void {
+  protected onUpdateRequest(_: Message): void {
     // Convert the value and page into percentages.
     let value = (this._value * 100) / this._maximum;
     let page = (this._page * 100) / (this._page + this._maximum);
diff --git a/packages/widgets/src/splitlayout.ts b/packages/widgets/src/splitlayout.ts
index 6b71501e..f87cc3c5 100644
--- a/packages/widgets/src/splitlayout.ts
+++ b/packages/widgets/src/splitlayout.ts
@@ -335,7 +335,7 @@ export class SplitLayout extends PanelLayout {
   protected moveWidget(
     fromIndex: number,
     toIndex: number,
-    widget: Widget
+    _: Widget
   ): void {
     // Move the item, sizer, and handle for the widget.
     ArrayExt.move(this._items, fromIndex, toIndex);
@@ -402,14 +402,14 @@ export class SplitLayout extends PanelLayout {
   /**
    * A message handler invoked on a `'child-shown'` message.
    */
-  protected onChildShown(msg: Widget.ChildMessage): void {
+  protected onChildShown(_: Widget.ChildMessage): void {
     this.parent!.fit();
   }
 
   /**
    * A message handler invoked on a `'child-hidden'` message.
    */
-  protected onChildHidden(msg: Widget.ChildMessage): void {
+  protected onChildHidden(_: Widget.ChildMessage): void {
     this.parent!.fit();
   }
 
@@ -425,7 +425,7 @@ export class SplitLayout extends PanelLayout {
   /**
    * A message handler invoked on an `'update-request'` message.
    */
-  protected onUpdateRequest(msg: Message): void {
+  protected onUpdateRequest(_: Message): void {
     if (this.parent!.isVisible) {
       this._update(-1, -1);
     }
@@ -434,7 +434,7 @@ export class SplitLayout extends PanelLayout {
   /**
    * A message handler invoked on a `'fit-request'` message.
    */
-  protected onFitRequest(msg: Message): void {
+  protected onFitRequest(_: Message): void {
     if (this.parent!.isAttached) {
       this._fit();
     }
@@ -808,7 +808,7 @@ namespace Private {
   export const stretchProperty = new AttachedProperty<Widget, number>({
     name: 'stretch',
     create: () => 0,
-    coerce: (owner, value) => Math.max(0, Math.floor(value)),
+    coerce: (_, value) => Math.max(0, Math.floor(value)),
     changed: onChildSizingChanged
   });
 
@@ -848,7 +848,7 @@ namespace Private {
       return [];
     }
     let sum = values.reduce((a, b) => a + Math.abs(b), 0);
-    return sum === 0 ? values.map(v => 1 / n) : values.map(v => v / sum);
+    return sum === 0 ? values.map(_ => 1 / n) : values.map(v => v / sum);
   }
 
   /**
diff --git a/packages/widgets/src/splitpanel.ts b/packages/widgets/src/splitpanel.ts
index a87508c3..0eb5ab0b 100644
--- a/packages/widgets/src/splitpanel.ts
+++ b/packages/widgets/src/splitpanel.ts
@@ -197,7 +197,7 @@ export class SplitPanel extends Panel {
   /**
    * A message handler invoked on a `'before-attach'` message.
    */
-  protected onBeforeAttach(msg: Message): void {
+  protected onBeforeAttach(_: Message): void {
     this.node.addEventListener('mousedown', this);
     this.node.addEventListener('pointerdown', this);
   }
@@ -205,7 +205,7 @@ export class SplitPanel extends Panel {
   /**
    * A message handler invoked on an `'after-detach'` message.
    */
-  protected onAfterDetach(msg: Message): void {
+  protected onAfterDetach(_: Message): void {
     this.node.removeEventListener('mousedown', this);
     this.node.removeEventListener('pointerdown', this);
     this._releaseMouse();
diff --git a/packages/widgets/src/stackedlayout.ts b/packages/widgets/src/stackedlayout.ts
index 36bdf535..be3b9f67 100644
--- a/packages/widgets/src/stackedlayout.ts
+++ b/packages/widgets/src/stackedlayout.ts
@@ -141,7 +141,7 @@ export class StackedLayout extends PanelLayout {
   protected moveWidget(
     fromIndex: number,
     toIndex: number,
-    widget: Widget
+    _: Widget
   ): void {
     // Move the layout item for the widget.
     ArrayExt.move(this._items, fromIndex, toIndex);
@@ -216,14 +216,14 @@ export class StackedLayout extends PanelLayout {
   /**
    * A message handler invoked on a `'child-shown'` message.
    */
-  protected onChildShown(msg: Widget.ChildMessage): void {
+  protected onChildShown(_: Widget.ChildMessage): void {
     this.parent!.fit();
   }
 
   /**
    * A message handler invoked on a `'child-hidden'` message.
    */
-  protected onChildHidden(msg: Widget.ChildMessage): void {
+  protected onChildHidden(_: Widget.ChildMessage): void {
     this.parent!.fit();
   }
 
@@ -239,7 +239,7 @@ export class StackedLayout extends PanelLayout {
   /**
    * A message handler invoked on an `'update-request'` message.
    */
-  protected onUpdateRequest(msg: Message): void {
+  protected onUpdateRequest(_: Message): void {
     if (this.parent!.isVisible) {
       this._update(-1, -1);
     }
@@ -248,7 +248,7 @@ export class StackedLayout extends PanelLayout {
   /**
    * A message handler invoked on a `'fit-request'` message.
    */
-  protected onFitRequest(msg: Message): void {
+  protected onFitRequest(_: Message): void {
     if (this.parent!.isAttached) {
       this._fit();
     }
diff --git a/packages/widgets/src/tabbar.ts b/packages/widgets/src/tabbar.ts
index 9f0e925a..8264e64c 100644
--- a/packages/widgets/src/tabbar.ts
+++ b/packages/widgets/src/tabbar.ts
@@ -612,7 +612,7 @@ export class TabBar<T> extends Widget {
   /**
    * A message handler invoked on a `'before-attach'` message.
    */
-  protected onBeforeAttach(msg: Message): void {
+  protected onBeforeAttach(_: Message): void {
     this.node.addEventListener('pointerdown', this);
     this.node.addEventListener('dblclick', this);
   }
@@ -620,7 +620,7 @@ export class TabBar<T> extends Widget {
   /**
    * A message handler invoked on an `'after-detach'` message.
    */
-  protected onAfterDetach(msg: Message): void {
+  protected onAfterDetach(_: Message): void {
     this.node.removeEventListener('pointerdown', this);
     this.node.removeEventListener('dblclick', this);
     this._releaseMouse();
@@ -629,7 +629,7 @@ export class TabBar<T> extends Widget {
   /**
    * A message handler invoked on an `'update-request'` message.
    */
-  protected onUpdateRequest(msg: Message): void {
+  protected onUpdateRequest(_: Message): void {
     let titles = this._titles;
     let renderer = this.renderer;
     let currentTitle = this.currentTitle;
@@ -1195,7 +1195,7 @@ export class TabBar<T> extends Widget {
   /**
    * Handle the `changed` signal of a title object.
    */
-  private _onTitleChanged(sender: Title<T>): void {
+  private _onTitleChanged(_: Title<T>): void {
     this.update();
   }
 
@@ -1588,7 +1588,7 @@ export namespace TabBar {
      *
      * @returns A virtual element representing the tab close icon.
      */
-    renderCloseIcon(data: IRenderData<any>): VirtualElement {
+    renderCloseIcon(_: IRenderData<any>): VirtualElement {
       return h.div({ className: 'lm-TabBar-tabCloseIcon' });
     }
 
@@ -1773,7 +1773,7 @@ namespace Private {
      *
      * This will be `null` if the drag is not active.
      */
-    contentRect: ClientRect | null;
+    contentRect: DOMRect | null;
 
     /**
      * The disposable to clean up the cursor override.
diff --git a/packages/widgets/src/tabpanel.ts b/packages/widgets/src/tabpanel.ts
index ce47049e..68286bb5 100644
--- a/packages/widgets/src/tabpanel.ts
+++ b/packages/widgets/src/tabpanel.ts
@@ -295,7 +295,7 @@ export class TabPanel extends Widget {
    * Handle the `currentChanged` signal from the tab bar.
    */
   private _onCurrentChanged(
-    sender: TabBar<Widget>,
+    _: TabBar<Widget>,
     args: TabBar.ICurrentChangedArgs<Widget>
   ): void {
     // Extract the previous and current title from the args.
@@ -332,7 +332,7 @@ export class TabPanel extends Widget {
   /**
    * Handle the `tabAddRequested` signal from the tab bar.
    */
-  private _onTabAddRequested(sender: TabBar<Widget>, args: void): void {
+  private _onTabAddRequested(sender: TabBar<Widget>, _: void): void {
     this._addRequested.emit(sender);
   }
 
@@ -340,7 +340,7 @@ export class TabPanel extends Widget {
    * Handle the `tabActivateRequested` signal from the tab bar.
    */
   private _onTabActivateRequested(
-    sender: TabBar<Widget>,
+    _: TabBar<Widget>,
     args: TabBar.ITabActivateRequestedArgs<Widget>
   ): void {
     args.title.owner.activate();
@@ -350,7 +350,7 @@ export class TabPanel extends Widget {
    * Handle the `tabCloseRequested` signal from the tab bar.
    */
   private _onTabCloseRequested(
-    sender: TabBar<Widget>,
+    _: TabBar<Widget>,
     args: TabBar.ITabCloseRequestedArgs<Widget>
   ): void {
     args.title.owner.close();
@@ -360,7 +360,7 @@ export class TabPanel extends Widget {
    * Handle the `tabMoved` signal from the tab bar.
    */
   private _onTabMoved(
-    sender: TabBar<Widget>,
+    _: TabBar<Widget>,
     args: TabBar.ITabMovedArgs<Widget>
   ): void {
     this.stackedPanel.insertWidget(args.toIndex, args.title.owner);
@@ -369,7 +369,7 @@ export class TabPanel extends Widget {
   /**
    * Handle the `widgetRemoved` signal from the stacked panel.
    */
-  private _onWidgetRemoved(sender: StackedPanel, widget: Widget): void {
+  private _onWidgetRemoved(_: StackedPanel, widget: Widget): void {
     widget.node.removeAttribute('role');
     widget.node.removeAttribute('aria-labelledby');
     this.tabBar.removeTab(widget.title);
diff --git a/packages/widgets/src/widget.ts b/packages/widgets/src/widget.ts
index 9f5ef825..64279865 100644
--- a/packages/widgets/src/widget.ts
+++ b/packages/widgets/src/widget.ts
@@ -1,4 +1,3 @@
-/* eslint-disable @typescript-eslint/no-empty-function */
 // Copyright (c) Jupyter Development Team.
 // Distributed under the terms of the Modified BSD License.
 /*-----------------------------------------------------------------------------
@@ -624,7 +623,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation unparents or detaches the widget.
    */
-  protected onCloseRequest(msg: Message): void {
+  protected onCloseRequest(_: Message): void {
     if (this.parent) {
       this.parent = null;
     } else if (this.isAttached) {
@@ -638,7 +637,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onResize(msg: Widget.ResizeMessage): void {}
+  protected onResize(_: Widget.ResizeMessage): void { return; }
 
   /**
    * A message handler invoked on an `'update-request'` message.
@@ -646,7 +645,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onUpdateRequest(msg: Message): void {}
+  protected onUpdateRequest(_: Message): void { return; }
 
   /**
    * A message handler invoked on a `'fit-request'` message.
@@ -654,7 +653,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onFitRequest(msg: Message): void {}
+  protected onFitRequest(_: Message): void { return; }
 
   /**
    * A message handler invoked on an `'activate-request'` message.
@@ -662,7 +661,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onActivateRequest(msg: Message): void {}
+  protected onActivateRequest(_: Message): void { return; }
 
   /**
    * A message handler invoked on a `'before-show'` message.
@@ -670,7 +669,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onBeforeShow(msg: Message): void {}
+  protected onBeforeShow(_: Message): void { return; }
 
   /**
    * A message handler invoked on an `'after-show'` message.
@@ -678,7 +677,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onAfterShow(msg: Message): void {}
+  protected onAfterShow(_: Message): void { return; }
 
   /**
    * A message handler invoked on a `'before-hide'` message.
@@ -686,7 +685,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onBeforeHide(msg: Message): void {}
+  protected onBeforeHide(_: Message): void { return; }
 
   /**
    * A message handler invoked on an `'after-hide'` message.
@@ -694,7 +693,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onAfterHide(msg: Message): void {}
+  protected onAfterHide(_: Message): void { return; }
 
   /**
    * A message handler invoked on a `'before-attach'` message.
@@ -702,7 +701,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onBeforeAttach(msg: Message): void {}
+  protected onBeforeAttach(_: Message): void { return; }
 
   /**
    * A message handler invoked on an `'after-attach'` message.
@@ -710,7 +709,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onAfterAttach(msg: Message): void {}
+  protected onAfterAttach(_: Message): void { return; }
 
   /**
    * A message handler invoked on a `'before-detach'` message.
@@ -718,7 +717,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onBeforeDetach(msg: Message): void {}
+  protected onBeforeDetach(_: Message): void { return; }
 
   /**
    * A message handler invoked on an `'after-detach'` message.
@@ -726,7 +725,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onAfterDetach(msg: Message): void {}
+  protected onAfterDetach(_: Message): void { return; }
 
   /**
    * A message handler invoked on a `'child-added'` message.
@@ -734,7 +733,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onChildAdded(msg: Widget.ChildMessage): void {}
+  protected onChildAdded(_: Message): void { return; }
 
   /**
    * A message handler invoked on a `'child-removed'` message.
@@ -742,7 +741,7 @@ export class Widget implements IMessageHandler, IObservableDisposable {
    * #### Notes
    * The default implementation of this handler is a no-op.
    */
-  protected onChildRemoved(msg: Widget.ChildMessage): void {}
+  protected onChildRemoved(_: Message): void { return; }
 
   private _flags = 0;
   private _layout: Layout | null = null;
diff --git a/packages/widgets/tests/src/accordionlayout.spec.ts b/packages/widgets/tests/src/accordionlayout.spec.ts
index 82ff663f..bfae4363 100644
--- a/packages/widgets/tests/src/accordionlayout.spec.ts
+++ b/packages/widgets/tests/src/accordionlayout.spec.ts
@@ -3,13 +3,13 @@
 
 import { each, every } from '@lumino/algorithm';
 import { Message } from '@lumino/messaging';
-import { AccordionLayout, Title, Widget } from '@lumino/widgets';
+import { AccordionLayout, Widget } from '@lumino/widgets';
 import { expect } from 'chai';
 
 const renderer: AccordionLayout.IRenderer = {
   titleClassName: '.lm-AccordionTitle',
   createHandle: () => document.createElement('div'),
-  createSectionTitle: (title: Title<Widget>) => document.createElement('h3')
+  createSectionTitle: () => document.createElement('h3')
 };
 
 class LogAccordionLayout extends AccordionLayout {
diff --git a/packages/widgets/tests/src/accordionpanel.spec.ts b/packages/widgets/tests/src/accordionpanel.spec.ts
index b515dc30..4168ab34 100644
--- a/packages/widgets/tests/src/accordionpanel.spec.ts
+++ b/packages/widgets/tests/src/accordionpanel.spec.ts
@@ -15,7 +15,7 @@ const bubbles = true;
 const renderer: AccordionPanel.IRenderer = {
   titleClassName: '.lm-AccordionTitle',
   createHandle: () => document.createElement('div'),
-  createSectionTitle: (title: Title<Widget>) => document.createElement('h3')
+  createSectionTitle: () => document.createElement('h3')
 };
 
 class LogAccordionPanel extends AccordionPanel {
diff --git a/packages/widgets/tests/src/contextmenu.spec.ts b/packages/widgets/tests/src/contextmenu.spec.ts
index d2a9753f..4f186754 100644
--- a/packages/widgets/tests/src/contextmenu.spec.ts
+++ b/packages/widgets/tests/src/contextmenu.spec.ts
@@ -12,8 +12,6 @@ import { expect } from 'chai';
 
 import { CommandRegistry } from '@lumino/commands';
 
-import { JSONObject } from '@lumino/coreutils';
-
 import { ContextMenu } from '@lumino/widgets';
 
 describe('@lumino/widgets', () => {
@@ -21,19 +19,19 @@ describe('@lumino/widgets', () => {
 
   before(() => {
     commands.addCommand('test-1', {
-      execute: (args: JSONObject) => {},
+      execute: () => undefined,
       label: 'Test 1 Label'
     });
     commands.addCommand('test-2', {
-      execute: (args: JSONObject) => {},
+      execute: () => undefined,
       label: 'Test 2 Label'
     });
     commands.addCommand('test-3', {
-      execute: (args: JSONObject) => {},
+      execute: () => undefined,
       label: 'Test 3 Label'
     });
     commands.addCommand('test-4', {
-      execute: (args: JSONObject) => {},
+      execute: () => undefined,
       label: 'Test 4 Label'
     });
   });
diff --git a/packages/widgets/tests/src/focustracker.spec.ts b/packages/widgets/tests/src/focustracker.spec.ts
index 702be459..bf8083cf 100644
--- a/packages/widgets/tests/src/focustracker.spec.ts
+++ b/packages/widgets/tests/src/focustracker.spec.ts
@@ -87,7 +87,7 @@ describe('@lumino/widgets', () => {
         tracker.add(widget1);
         focusWidget(widget0);
         let emitArgs: FocusTracker.IChangedArgs<Widget> | null = null;
-        tracker.currentChanged.connect((sender, args) => {
+        tracker.currentChanged.connect((_, args) => {
           emitArgs = args;
         });
         focusWidget(widget1);
@@ -102,7 +102,7 @@ describe('@lumino/widgets', () => {
         focusWidget(widget);
         tracker.add(widget);
         let emitArgs: FocusTracker.IChangedArgs<Widget> | null = null;
-        tracker.currentChanged.connect((sender, args) => {
+        tracker.currentChanged.connect((_, args) => {
           emitArgs = args;
         });
         blurWidget(widget);
@@ -120,7 +120,7 @@ describe('@lumino/widgets', () => {
         tracker.add(widget1);
         focusWidget(widget0);
         let emitArgs: FocusTracker.IChangedArgs<Widget> | null = null;
-        tracker.activeChanged.connect((sender, args) => {
+        tracker.activeChanged.connect((_, args) => {
           emitArgs = args;
         });
         focusWidget(widget1);
@@ -135,7 +135,7 @@ describe('@lumino/widgets', () => {
         focusWidget(widget);
         tracker.add(widget);
         let emitArgs: FocusTracker.IChangedArgs<Widget> | null = null;
-        tracker.activeChanged.connect((sender, args) => {
+        tracker.activeChanged.connect((_, args) => {
           emitArgs = args;
         });
         blurWidget(widget);
diff --git a/packages/widgets/tests/src/menu.spec.ts b/packages/widgets/tests/src/menu.spec.ts
index b7362c70..ac11b28e 100644
--- a/packages/widgets/tests/src/menu.spec.ts
+++ b/packages/widgets/tests/src/menu.spec.ts
@@ -11,8 +11,6 @@ import { expect } from 'chai';
 
 import { CommandRegistry } from '@lumino/commands';
 
-import { JSONObject } from '@lumino/coreutils';
-
 import { Platform } from '@lumino/domutils';
 
 import { Message } from '@lumino/messaging';
@@ -66,7 +64,7 @@ describe('@lumino/widgets', () => {
   let executed = '';
   const iconClass = 'foo';
   const iconRenderer = {
-    render: (host: HTMLElement, options?: any) => {
+    render: (host: HTMLElement) => {
       const renderNode = document.createElement('div');
       host.classList.add(iconClass);
       host.appendChild(renderNode);
@@ -75,7 +73,7 @@ describe('@lumino/widgets', () => {
 
   before(() => {
     commands.addCommand('test', {
-      execute: (args: JSONObject) => {
+      execute: () => {
         executed = 'test';
       },
       label: 'Test Label',
@@ -86,36 +84,36 @@ describe('@lumino/widgets', () => {
       mnemonic: 0
     });
     commands.addCommand('test-toggled', {
-      execute: (args: JSONObject) => {
+      execute: () => {
         executed = 'test-toggled';
       },
       label: 'Test Toggled Label',
       icon: iconRenderer,
       className: 'testClass',
-      isToggled: (args: JSONObject) => true,
+      isToggled: () => true,
       mnemonic: 6
     });
     commands.addCommand('test-disabled', {
-      execute: (args: JSONObject) => {
+      execute: () => {
         executed = 'test-disabled';
       },
       label: 'Test Disabled Label',
       icon: iconRenderer,
       className: 'testClass',
-      isEnabled: (args: JSONObject) => false,
+      isEnabled: () => false,
       mnemonic: 5
     });
     commands.addCommand('test-hidden', {
-      execute: (args: JSONObject) => {
+      execute: () => {
         executed = 'test-hidden';
       },
       label: 'Hidden Label',
       icon: iconRenderer,
       className: 'testClass',
-      isVisible: (args: JSONObject) => false
+      isVisible: () => false
     });
     commands.addCommand('test-zenith', {
-      execute: (args: JSONObject) => {
+      execute: () => {
         executed = 'test-zenith';
       },
       label: 'Zenith Label',
@@ -167,7 +165,7 @@ describe('@lumino/widgets', () => {
       it('should be emitted just before the menu is closed', () => {
         let called = false;
         menu.open(0, 0);
-        menu.aboutToClose.connect((sender, args) => {
+        menu.aboutToClose.connect(() => {
           called = true;
         });
         menu.close();
@@ -190,7 +188,7 @@ describe('@lumino/widgets', () => {
       it('should be emitted when a left arrow key is pressed and a submenu cannot be opened or closed', () => {
         let called = false;
         menu.open(0, 0);
-        menu.menuRequested.connect((sender, args) => {
+        menu.menuRequested.connect((_, args) => {
           expect(args).to.equal('previous');
           called = true;
         });
@@ -206,7 +204,7 @@ describe('@lumino/widgets', () => {
       it('should be emitted when a right arrow key is pressed and a submenu cannot be opened or closed', () => {
         let called = false;
         menu.open(0, 0);
-        menu.menuRequested.connect((sender, args) => {
+        menu.menuRequested.connect((_, args) => {
           expect(args).to.equal('next');
           called = true;
         });
@@ -227,7 +225,7 @@ describe('@lumino/widgets', () => {
         menu.triggerActiveItem();
         let called = false;
         let submenuCalled = false;
-        menu.menuRequested.connect((sender, args) => {
+        menu.menuRequested.connect((_, args) => {
           expect(args).to.equal('next');
           called = true;
         });
diff --git a/packages/widgets/tests/src/menubar.spec.ts b/packages/widgets/tests/src/menubar.spec.ts
index 84845b4a..af42a6df 100644
--- a/packages/widgets/tests/src/menubar.spec.ts
+++ b/packages/widgets/tests/src/menubar.spec.ts
@@ -77,7 +77,7 @@ describe('@lumino/widgets', () => {
   before(() => {
     commands = new CommandRegistry();
     const iconRenderer = {
-      render: (host: HTMLElement, options?: any) => {
+      render: (host: HTMLElement) => {
         const renderNode = document.createElement('div');
         renderNode.className = 'foo';
         host.appendChild(renderNode);
@@ -90,9 +90,7 @@ describe('@lumino/widgets', () => {
       label: 'LABEL',
       icon: iconRenderer,
       className: 'bar',
-      isToggled: (args: JSONObject) => {
-        return true;
-      },
+      isToggled: () => true,
       mnemonic: 1
     });
     let kbd = commands.addKeyBinding({
diff --git a/packages/widgets/tests/src/panellayout.spec.ts b/packages/widgets/tests/src/panellayout.spec.ts
index cae74616..1f1b465f 100644
--- a/packages/widgets/tests/src/panellayout.spec.ts
+++ b/packages/widgets/tests/src/panellayout.spec.ts
@@ -23,7 +23,7 @@ import { PanelLayout, Widget } from '@lumino/widgets';
 class LogHook implements IMessageHook {
   messages: string[] = [];
 
-  messageHook(target: IMessageHandler, msg: Message): boolean {
+  messageHook(_: IMessageHandler, msg: Message): boolean {
     this.messages.push(msg.type);
     return true;
   }
diff --git a/packages/widgets/tests/src/splitlayout.spec.ts b/packages/widgets/tests/src/splitlayout.spec.ts
index a899933f..e707bcf2 100644
--- a/packages/widgets/tests/src/splitlayout.spec.ts
+++ b/packages/widgets/tests/src/splitlayout.spec.ts
@@ -90,7 +90,7 @@ class LogSplitLayout extends SplitLayout {
 class LogHook implements IMessageHook {
   messages: string[] = [];
 
-  messageHook(target: IMessageHandler, msg: Message): boolean {
+  messageHook(_: IMessageHandler, msg: Message): boolean {
     this.messages.push(msg.type);
     return true;
   }
diff --git a/packages/widgets/tests/src/stackedlayout.spec.ts b/packages/widgets/tests/src/stackedlayout.spec.ts
index 612e092b..dfcb8fd1 100644
--- a/packages/widgets/tests/src/stackedlayout.spec.ts
+++ b/packages/widgets/tests/src/stackedlayout.spec.ts
@@ -23,7 +23,7 @@ import { StackedLayout, Widget } from '@lumino/widgets';
 class LogHook implements IMessageHook {
   messages: string[] = [];
 
-  messageHook(target: IMessageHandler, msg: Message): boolean {
+  messageHook(_: IMessageHandler, msg: Message): boolean {
     this.messages.push(msg.type);
     return true;
   }
diff --git a/packages/widgets/tests/src/tabbar.spec.ts b/packages/widgets/tests/src/tabbar.spec.ts
index 082eb979..aec7a2e5 100644
--- a/packages/widgets/tests/src/tabbar.spec.ts
+++ b/packages/widgets/tests/src/tabbar.spec.ts
@@ -76,7 +76,7 @@ function startDrag(
   // Force an update.
   MessageLoop.sendMessage(bar, Widget.Msg.UpdateRequest);
   let called = false;
-  bar.tabDetachRequested.connect((sender, args) => {
+  bar.tabDetachRequested.connect(() => {
     called = true;
   });
   let rect = bar.contentNode.getBoundingClientRect();
@@ -189,7 +189,7 @@ describe('@lumino/widgets', () => {
       it('should not be emitted when another tab is inserted', () => {
         populateBar(bar);
         let called = false;
-        bar.currentChanged.connect((sender, args) => {
+        bar.currentChanged.connect(() => {
           called = true;
         });
         let widget = new Widget();
@@ -201,7 +201,7 @@ describe('@lumino/widgets', () => {
         populateBar(bar);
         let called = false;
         bar.currentIndex = 1;
-        bar.currentChanged.connect((sender, args) => {
+        bar.currentChanged.connect(() => {
           called = true;
         });
         bar.removeTab(bar.titles[0]);
@@ -211,7 +211,7 @@ describe('@lumino/widgets', () => {
       it('should not be emitted when the current tab is moved', () => {
         populateBar(bar);
         let called = false;
-        bar.currentChanged.connect((sender, args) => {
+        bar.currentChanged.connect(() => {
           called = true;
         });
         bar.insertTab(2, bar.titles[0]);
@@ -255,7 +255,7 @@ describe('@lumino/widgets', () => {
       it('should not be emitted when a tab is moved programmatically', () => {
         populateBar(bar);
         let called = false;
-        bar.tabMoved.connect((sender, args) => {
+        bar.tabMoved.connect(() => {
           called = true;
         });
         bar.insertTab(2, bar.titles[0]);
@@ -531,7 +531,7 @@ describe('@lumino/widgets', () => {
       it('should be handled by calling `releaseMouse` and removing the tab', () => {
         simulateOnNode(tab, 'pointerdown');
         let called = false;
-        bar.tabDetachRequested.connect((sender, args) => {
+        bar.tabDetachRequested.connect((_, args) => {
           bar.releaseMouse();
           bar.removeTabAt(args.index);
           called = true;
@@ -550,7 +550,7 @@ describe('@lumino/widgets', () => {
       it('should only be emitted once per drag cycle', () => {
         simulateOnNode(tab, 'pointerdown');
         let called = 0;
-        bar.tabDetachRequested.connect((sender, args) => {
+        bar.tabDetachRequested.connect((_, args) => {
           bar.releaseMouse();
           bar.removeTabAt(args.index);
           called++;
@@ -577,7 +577,7 @@ describe('@lumino/widgets', () => {
       it('should add the `lm-mod-dragging` class to the tab and the bar', () => {
         simulateOnNode(tab, 'pointerdown');
         let called = false;
-        bar.tabDetachRequested.connect((sender, args) => {
+        bar.tabDetachRequested.connect(() => {
           expect(tab.classList.contains('lm-mod-dragging')).to.equal(true);
           expect(bar.hasClass('lm-mod-dragging')).to.equal(true);
           called = true;
@@ -1059,7 +1059,7 @@ describe('@lumino/widgets', () => {
         populateBar(bar);
         let called = false;
         bar.currentIndex = -1;
-        bar.currentChanged.connect((sender, args) => {
+        bar.currentChanged.connect(() => {
           called = true;
         });
         bar.clearTabs();
@@ -1122,7 +1122,7 @@ describe('@lumino/widgets', () => {
           startDrag(bar, 1, 'up');
           let called = false;
           let rect = closeIcon.getBoundingClientRect();
-          bar.tabCloseRequested.connect((sender, args) => {
+          bar.tabCloseRequested.connect(() => {
             called = true;
           });
           closeIcon.dispatchEvent(
@@ -1147,7 +1147,7 @@ describe('@lumino/widgets', () => {
         it('should do nothing if the click is not on a close icon', () => {
           let called = false;
           let rect = closeIcon.getBoundingClientRect();
-          bar.tabCloseRequested.connect((sender, args) => {
+          bar.tabCloseRequested.connect(() => {
             called = true;
           });
           closeIcon.dispatchEvent(
@@ -1174,7 +1174,7 @@ describe('@lumino/widgets', () => {
           let called = false;
           bar.titles[0].closable = false;
           let rect = closeIcon.getBoundingClientRect();
-          bar.tabCloseRequested.connect((sender, args) => {
+          bar.tabCloseRequested.connect(() => {
             called = true;
           });
           closeIcon.dispatchEvent(
@@ -1230,7 +1230,7 @@ describe('@lumino/widgets', () => {
           startDrag(bar, 1, 'up');
           let called = false;
           let rect = tab.getBoundingClientRect();
-          bar.tabCloseRequested.connect((sender, args) => {
+          bar.tabCloseRequested.connect(() => {
             called = true;
           });
           tab.dispatchEvent(
@@ -1255,7 +1255,7 @@ describe('@lumino/widgets', () => {
         it('should do nothing if the click is not on the tab', () => {
           let called = false;
           let rect = tab.getBoundingClientRect();
-          bar.tabCloseRequested.connect((sender, args) => {
+          bar.tabCloseRequested.connect(() => {
             called = true;
           });
           tab.dispatchEvent(
@@ -1282,7 +1282,7 @@ describe('@lumino/widgets', () => {
           let called = false;
           bar.titles[0].closable = false;
           let rect = tab.getBoundingClientRect();
-          bar.tabCloseRequested.connect((sender, args) => {
+          bar.tabCloseRequested.connect(() => {
             called = true;
           });
           tab.dispatchEvent(
@@ -1369,7 +1369,7 @@ describe('@lumino/widgets', () => {
         it('should do nothing if there is a drag in progress', () => {
           simulateOnNode(tab, 'pointerdown');
           let called = 0;
-          bar.tabDetachRequested.connect((sender, args) => {
+          bar.tabDetachRequested.connect(() => {
             called++;
           });
           let rect = bar.contentNode.getBoundingClientRect();
@@ -1394,7 +1394,7 @@ describe('@lumino/widgets', () => {
         it('should bail if the drag threshold is not exceeded', () => {
           simulateOnNode(tab, 'pointerdown');
           let called = false;
-          bar.tabDetachRequested.connect((sender, args) => {
+          bar.tabDetachRequested.connect(() => {
             bar.releaseMouse();
             called = true;
           });
@@ -1434,7 +1434,7 @@ describe('@lumino/widgets', () => {
         it('should bail if the signal handler aborted the drag', () => {
           simulateOnNode(tab, 'pointerdown');
           let called = false;
-          bar.tabDetachRequested.connect((sender, args) => {
+          bar.tabDetachRequested.connect(() => {
             bar.releaseMouse();
             called = true;
           });
@@ -1455,7 +1455,7 @@ describe('@lumino/widgets', () => {
         it('should update the positions of the tabs', () => {
           simulateOnNode(tab, 'pointerdown');
           let called = false;
-          bar.tabDetachRequested.connect((sender, args) => {
+          bar.tabDetachRequested.connect(() => {
             called = true;
           });
           let rect = bar.contentNode.getBoundingClientRect();
diff --git a/packages/widgets/tests/src/tabpanel.spec.ts b/packages/widgets/tests/src/tabpanel.spec.ts
index 12441594..e2aebc38 100644
--- a/packages/widgets/tests/src/tabpanel.spec.ts
+++ b/packages/widgets/tests/src/tabpanel.spec.ts
@@ -75,7 +75,7 @@ describe('@lumino/widgets', () => {
         panel.addWidget(new Widget());
         panel.addWidget(new Widget());
         let called = false;
-        panel.currentChanged.connect((sender, args) => {
+        panel.currentChanged.connect(() => {
           called = true;
         });
         panel.insertWidget(0, new Widget());
@@ -88,7 +88,7 @@ describe('@lumino/widgets', () => {
         panel.addWidget(new Widget());
         let called = false;
         panel.currentIndex = 1;
-        panel.currentChanged.connect((sender, args) => {
+        panel.currentChanged.connect(() => {
           called = true;
         });
         panel.widgets[0].parent = null;
@@ -100,7 +100,7 @@ describe('@lumino/widgets', () => {
         panel.addWidget(new Widget());
         panel.addWidget(new Widget());
         let called = false;
-        panel.currentChanged.connect((sender, args) => {
+        panel.currentChanged.connect(() => {
           called = true;
         });
         panel.insertWidget(2, panel.widgets[0]);
@@ -254,7 +254,7 @@ describe('@lumino/widgets', () => {
           w.node.tabIndex = -1;
         });
         Widget.attach(panel, document.body);
-        panel.tabBar.currentChanged.connect((sender, args) => {
+        panel.tabBar.currentChanged.connect((_, args) => {
           expect(widgets[args.previousIndex].isVisible).to.equal(false);
           expect(widgets[args.currentIndex].isVisible).to.equal(true);
         });
diff --git a/packages/widgets/tests/src/title.spec.ts b/packages/widgets/tests/src/title.spec.ts
index f4f03ef6..22b355f8 100644
--- a/packages/widgets/tests/src/title.spec.ts
+++ b/packages/widgets/tests/src/title.spec.ts
@@ -40,7 +40,7 @@ describe('@lumino/widgets', () => {
       it('should be cleared if it is disposed', () => {
         let called = false;
         let title = new Title({ owner });
-        title.changed.connect((sender, arg) => {
+        title.changed.connect(() => {
           called = true;
         });
 
@@ -90,7 +90,7 @@ describe('@lumino/widgets', () => {
       it('should not emit the changed signal when the value does not change', () => {
         let called = false;
         let title = new Title({ owner, label: 'foo' });
-        title.changed.connect((sender, arg) => {
+        title.changed.connect(() => {
           called = true;
         });
         title.label = 'foo';
@@ -130,7 +130,7 @@ describe('@lumino/widgets', () => {
       it('should not emit the changed signal when the value does not change', () => {
         let called = false;
         let title = new Title({ owner, mnemonic: 1 });
-        title.changed.connect((sender, arg) => {
+        title.changed.connect(() => {
           called = true;
         });
         title.mnemonic = 1;
@@ -140,7 +140,7 @@ describe('@lumino/widgets', () => {
 
     describe('#icon', () => {
       const iconRenderer = {
-        render: (host: HTMLElement, options?: any) => {
+        render: (host: HTMLElement) => {
           const renderNode = document.createElement('div');
           renderNode.className = 'foo';
           host.appendChild(renderNode);
@@ -179,7 +179,7 @@ describe('@lumino/widgets', () => {
       it('should not emit the changed signal when the value does not change', () => {
         let called = false;
         let title = new Title({ owner, icon: iconRenderer });
-        title.changed.connect((sender, arg) => {
+        title.changed.connect(() => {
           called = true;
         });
         title.icon = iconRenderer;
@@ -219,7 +219,7 @@ describe('@lumino/widgets', () => {
       it('should not emit the changed signal when the value does not change', () => {
         let called = false;
         let title = new Title({ owner, caption: 'foo' });
-        title.changed.connect((sender, arg) => {
+        title.changed.connect(() => {
           called = true;
         });
         title.caption = 'foo';
@@ -259,7 +259,7 @@ describe('@lumino/widgets', () => {
       it('should not emit the changed signal when the value does not change', () => {
         let called = false;
         let title = new Title({ owner, className: 'foo' });
-        title.changed.connect((sender, arg) => {
+        title.changed.connect(() => {
           called = true;
         });
         title.className = 'foo';
@@ -299,7 +299,7 @@ describe('@lumino/widgets', () => {
       it('should not emit the changed signal when the value does not change', () => {
         let called = false;
         let title = new Title({ owner, closable: false });
-        title.changed.connect((sender, arg) => {
+        title.changed.connect(() => {
           called = true;
         });
         title.closable = false;
@@ -340,7 +340,7 @@ describe('@lumino/widgets', () => {
         let called = false;
         let dataset = { foo: '12' };
         let title = new Title({ owner, dataset });
-        title.changed.connect((sender, arg) => {
+        title.changed.connect(() => {
           called = true;
         });
         title.dataset = dataset;
diff --git a/tsconfigbase.json b/tsconfigbase.json
index 1d3a2bf6..2cc42caf 100644
--- a/tsconfigbase.json
+++ b/tsconfigbase.json
@@ -14,6 +14,7 @@
     "noEmitOnError": true,
     "noImplicitAny": true,
     "noUnusedLocals": true,
+    "noUnusedParameters": true,
     "preserveWatchOutput": true,
     "resolveJsonModule": true,
     "sourceMap": true,
```

Reply:
Thanks for trying.

These are indeed numerous changes. I let you decide what you think is the best option about `noUnusedParameters`. It may indeed be too annoying.

Reply:
I propose we leave this parameter turned off.